Tests: Storage. Покрытие API модуля тестами (LocalFS)

dr-wh0 opened the pull request:
# Описание изменений

Реализованы **интеграционные тесты** API модуля хранилищ (_**Storage**_) для всех эндпоинтов работы с хранилищами (и локальными дисками) в **LocalFS**: включая создание, удаление и управление хранилищами и разделами дисков (партициями). **Тесты NFS-хранилищ на данный момент не реализованы.**

В качестве инфраструктуры для тестов добавлены специализированные фикстуры и утилиы очистки. Тесты используют реальную файловую систему. Тесты покрывают основные сценарии работы, включая различную валидацию данных и обработку ошибок.

# Ключевые изменения

**Конфигурация тестовой среды:**
- Используется уже используемая тестами других модулей переменная окружения, которая указывается в конфигурации `.env.test` перед тестами:
```
	TEST_STORAGE_PATH=/dev/vdb
```

**Тестовая инфраструктура:**
- Фикстуры в `conftest.py`:
	- `cleanup_storages` - очищает все хранилища до и после теста
	- `target_disk_path` - путь к тестовому диску с удалением созданных тестовых партиций после теста
	- `local_partition` - создает тестовый раздел (партицию) на диске
	- `storage` - создает тестовое хранилище (localfs) и удаляет его после теста

-  Утилиты очистки:
	- `cleanup_all_storages()` - удаляет все хранилища из базы данных
	- `cleanup_all_volumes()` и `cleanup_all_templates()` - очистка связанных ресурсов
	- `get_disk_partitions()` - предоставляет список партиций диска
	- `cleanup_partitions()` - удаляет список тестовых партиций

**Реализованные тесты:**

*Создание хранилищ и партиций (localfs):*
- `test_create_storage_localfs_success` - успешное создание LocalFS-хранилища
- `test_create_storage_missing_name` - создание без имени
- `test_create_storage_missing_storage_type` - создание без типа хранилища
- `test_create_storage_missing_specs` - создание без спецификаций
- `test_create_storage_invalid_name` - создание с невалидным именем
- `test_create_storage_invalid_fs_type` - создание с невалидным типом файловой системы
- `test_create_storage_invalid_description_chars` - создание с невалидным описанием
- `test_create_storage_nonexistent_path` - создание с несуществующим путем
- `test_create_storage_duplicate_path` - создание с дублирующимся путем
- `test_create_storage_unauthorized` - проверка создания хранилиза без авторизации
- `test_create_local_partition_success` - успешное создание раздела диска
- `test_create_local_partition_nonexistent_disk` - создание на несуществующем диске
- `test_create_local_partition_insufficient_space` - создание с недостаточным местом
- `test_create_local_partition_unauthorized` - проверка создания раздела диска без авторизации

*Удаление:*
- `test_delete_storage_success` - успешное удаление хранилища
- `test_delete_storage_invalid_uuid` - удаление с невалидным UUID хранилища
- `test_delete_storage_not_found` - удаление несуществующего хранилища
- `test_delete_storage_with_attached_volume` - удаление хранилища с прикрепленными томами
- `test_delete_storage_with_attached_template` - удаление хранилища с прикрепленными шаблонами
- `test_delete_storage_unauthorized` - проверка удаления хранилища без авторизации
- `test_delete_local_partition_success` - успешное удаление раздела диска
- `test_delete_local_partition_nonexistent` - удаление несуществующего раздела диска
- `test_delete_local_partition_unauthorized` - проверка удаления раздела без авторизации

*Получение данных (GET):*
- `test_get_storages_success` - получение списка хранилищ
- `test_get_storages_empty_success` - получение пустого списка хранилищ
- `test_get_storage_success` - получение хранилища по ID
- `test_get_storage_nonexistent` - получение несуществующего хранилища
- `test_get_storage_invalid_id` - получение с невалидным ID
- `test_get_local_disks_success` - получение списка локальных дисков
- `test_get_local_disk_partitions_success` - получение информации о разделах
- `test_get_local_disk_partitions_invalid_disk` - получение с невалидным диском
- `test_get_local_disk_partitions_missing_param` - получение без параметров
- Все GET-ручки имеют соответствующие тесты `unauthorized`-доступа


**Общая структура файлов для тестов:**

```
tests/
├── conftest.py  # Фикстуры для тестов
├── api/
│    └── localfs/
│        ├── test_create.py  # POST /storages/create/, /storages/local-disks/create_partition/
│        ├── test_delete.py  # DELETE /storages/{id}/delete/, /storages/local-disks/delete_partition/
│        └── test_retrive.py  # GET /storages/, /storages/{id}/, /storages/local-disks/
```


# Особенности

- **На данный момент тесты работают только с LocalFS хранилищами.**
- Для тестов локальных дисков рекомендуется использовать выделенный тестовый диск со свободным разделом.
**Важно: для корректной работы с партициями диск должен быть не NVMe, рекомендуется иметь разметку GPT (GUID Partition Table).**

# Критерии приёмки

- [x] Сформирована структура файлов для тестов
- [x] Реализованы все тестовые сценарии для LocalFS
- [x] Добавлены необходимые фикстуры и утилиты очистки
- [x] Покрыты основные ошибки и edge cases
- [x] Тесты используют реальную инфраструктуру (FS, БД)
- [x] Добавлена [документация](https://github.com/Aerodisk/openvair-docs/pull/18) к тестам
- [x] Обеспечено стабильное прохождение всех тестов:
<img width="319" height="43" alt="image" src="https://github.com/user-attachments/assets/7f53f8af-87fe-4387-8845-67dce9c061a0" />


# Связанные задачи

- Issue #252

dr-wh0 committed:
Add storage create (localfs) API tests in Storage module

dr-wh0 committed:
Add storage delete API tests in Storage module

dr-wh0 committed:
Add local disk partitions operations API tests in Storage module

dr-wh0 committed:
Refactor cleanup_all_storages() in test utils

dr-wh0 committed:
Add test_create_nfs in storage module, ruff format

dr-wh0 committed:
Refactor files structure

dr-wh0 committed:
Delete nfs creation tests (temporary)

dr-wh0 committed:
Add partitions utils and refactor storage localfs tests

dr-wh0 committed:
Add check partitions creation and deletion in parted adapter in storage tests

dr-wh0 committed:
Merge branch 'dev' into tests/storage

dr-wh0 committed:
Merge branch 'dev' into tests/storage

dr-wh0 committed:
Refactor test utils - sort imports after merge with dev

dr-wh0 committed:
Refactor test files names: test_retrive/test_get -> test_retrieve

VladimirTrue reviewed (commented):
В целом ок, но нужны точечные исправления и/или аргументация решений

dr-wh0 commented on openvair/conftest.py:
```diff
@@ -136,9 +140,10 @@ def configure_pagination() -> None:
 
 
-@pytest.fixture(scope='module')
+@pytest.fixture(scope='function')
```
Нет, не поломало - исправил буквально 1 строчку в тестах Volume, теперь всё точно работает. Оно могло их немного замедлить, но это абсолютно не заметно и не критично для нас сейчас. 

Это необходимо было поменять, чтобы не писать отдельную фикстуру для тестов Storage.

dr-wh0 committed:
Merge branch 'refs/heads/dev' into tests/storage

dr-wh0 committed:
Fix fixture order in test_create_volume_unauthorized()

dr-wh0 committed:
Fix test file name: test_retrive.py -> test_retrieve.py

dr-wh0 commented on openvair/modules/network/tests/api/test_retrieve.py:
Да, переименование сделано в #266, уберу здесь

dr-wh0 committed:
Merge remote-tracking branch 'origin/tests/storage' into tests/storage

dr-wh0 committed:
Change back test files names

dr-wh0 commented on openvair/modules/storage/tests/api/localfs/test_retrieve.py:
```diff
@@ -0,0 +1,161 @@
+
+
+def test_get_storages_empty_success(
+    client: TestClient, cleanup_storages: None
```
Это была моя первая мысль, однако проблема заключается в том, что полагаясь лишь на `autouse=True` мы не можем контролировать порядок выполнения фикстур. Если для `test_retrieve` это не имеет особого значения, то для других тестов (delete/create) это может быть критично. Однако я попробую с использованием `autouse=True`, возможно здесь это удастся реализовать.

dr-wh0 replied:
Хмм, похоже для этих тестов это действительно работает верно, без проблем. Возможно, сталкивался с нарушением порядка выполнения в тестах NFS. 

Убрал ручное использование, использовал `autouse=True`.

dr-wh0 commented on openvair/modules/storage/tests/api/localfs/test_delete.py:
```diff
@@ -0,0 +1,153 @@
+    non_existent_id = str(uuid4())
+    response = client.delete(f'/storages/{non_existent_id}/delete')
+    assert response.status_code == status.HTTP_500_INTERNAL_SERVER_ERROR
+    assert 'not found' in response.text.lower()
```
Понимаю, но да, у нас мало вариантов. Как вариант, могу заменить на проверку на "EntityNotFoundError" (название самой ошибки). 

Но вообще, насколько я понимаю, конкретно в **этом** случае "not found" идёт напрямую из SqlAlchemy и не будет меняться. 

А вот ниже, в `test_delete_storage_with_attached_volume` и `test_delete_storage_with_attached_template` можно оставить только проверки на названия ошибок, например:
```python
response_text = response.text.lower()
assert 'storagehasobjects' in response_text
```

VladimirTrue replied:
Давай пока оставим так. В перспективе ещё предстоит подумать над тем как корректно сделать иерархию ошибок и как их ловить на самом верхнем уровне на границе API

dr-wh0 committed:
Refactor after review

- Add wait_full_deleting() check in test_delete_storage_success
- Remove model validation from test_get_local_disks_success
- Remove ruff noqa ARG001 from test files
- Change cleanup_storages() fixture to autouse=True

VladimirTrue commented:
approve! thumbsup

dr-wh0 committed:
Merge branch 'dev' into tests/storage

dr-wh0 committed:
Merge branch 'dev' into tests/storage

dr-wh0 committed:
Fix ruff and imports after merge
